crates/data_components/src/dynamodb/provider.rs

```
@@ -41,181 +44,123 @@ use datafusion::{
-                // Need to check if it's a valid JSON number
-                serde_json::Number::from_f64(f)
-                    .map(Value::Number)
-                    .unwrap_or(Value::String(n.clone()))
-            } else {
-                Value::String(n.clone())
+        for key in key_schema {
+            match key.key_type() {
+                KeyType::Hash => {
+                    partition_key = Some(key.attribute_name().to_string());
+                }
+                KeyType::Range => {
+                    sort_key = Some(key.attribute_name().to_string());
+                }
+                _ => {}
             }
         }
-        AttributeValue::Bool(b) => Value::Bool(*b),
-        AttributeValue::L(list) => Value::Array(list.iter().map(attribute_value_to_json).collect()),
-        AttributeValue::M(map) => attribute_map_to_json(map),
-        AttributeValue::Null(_) | _ => Value::Null,
-    }
-}
 
-/// Creates a projection expression for a `DynamoDB` scan request based on the provided schema and projection indices.
-/// Because projection expressions may use reserved words in `DynamoDB`, this function automatically generates expression attribute names for each column to avoid conflicts.
-/// The expression format used is `#c{idx}` for each projected column, where `{idx}` is the column projection index.
-/// See: <https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.ExpressionAttributeNames.html#Expressions.ExpressionAttributeNames.ReservedWords>
-/// Returns a tuple of (`projection_expression`, `expression_attribute_names`) for the `DynamoDB` scan.
-/// Returns None if no projection is required.
-fn projection_expression(
-    projection: Option<&Vec<usize>>,
-    schema: &SchemaRef,
-) -> Option<(String, HashMap<String, String>)> {
-    let projection = projection?;
-    if projection.is_empty() {
-        return None;
-    }
+        let Some(partition_key) = partition_key else {
```

Better to return an error here - this is dependent on an external service's behavior, not an invariant that we control.

---

An attacker could for example send a malformed payload pretending to be a dynamodb service and it would trigger Spice to panic, which is unacceptable